embassy-usb-hid/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+#![no_std]
+#![feature(generic_associated_types)]
+#![feature(type_alias_impl_trait)]
+
+//! Implements HID functionality for a usb-device device.
+
+// This mod MUST go first, so that the others see its macros.
+pub(crate) mod fmt;
+
+use core::mem::MaybeUninit;
+
+use embassy::channel::signal::Signal;
+use embassy::time::Duration;
+use embassy_usb::driver::{EndpointOut, ReadError};
+use embassy_usb::{
+    control::{ControlHandler, InResponse, OutResponse, Request, RequestType},
+    driver::{Driver, Endpoint, EndpointIn, WriteError},
+    UsbDeviceBuilder,
+};
+use futures_util::future::{select, Either};
+use futures_util::pin_mut;
+#[cfg(feature = "usbd-hid")]
+use ssmarshal::serialize;
+#[cfg(feature = "usbd-hid")]
+use usbd_hid::descriptor::AsInputReport;
+
+const USB_CLASS_HID: u8 = 0x03;
+const USB_SUBCLASS_NONE: u8 = 0x00;
+const USB_PROTOCOL_NONE: u8 = 0x00;
+
+// HID
+const HID_DESC_DESCTYPE_HID: u8 = 0x21;
+const HID_DESC_DESCTYPE_HID_REPORT: u8 = 0x22;
+const HID_DESC_SPEC_1_10: [u8; 2] = [0x10, 0x01];
+const HID_DESC_COUNTRY_UNSPEC: u8 = 0x00;
+
+const HID_REQ_SET_IDLE: u8 = 0x0a;
+const HID_REQ_GET_IDLE: u8 = 0x02;
+const HID_REQ_GET_REPORT: u8 = 0x01;
+const HID_REQ_SET_REPORT: u8 = 0x09;
+const HID_REQ_GET_PROTOCOL: u8 = 0x03;
+const HID_REQ_SET_PROTOCOL: u8 = 0x0b;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ReportId {
+    In(u8),
+    Out(u8),
+    Feature(u8),
+}
+
+impl ReportId {
+    fn try_from(value: u16) -> Result<Self, ()> {
+        match value >> 8 {
+            1 => Ok(ReportId::In(value as u8)),
+            2 => Ok(ReportId::Out(value as u8)),
+            3 => Ok(ReportId::Feature(value as u8)),
+            _ => Err(()),
+        }
+    }
+}
+
+pub struct State<'a, const IN_N: usize, const OUT_N: usize> {
+    control: MaybeUninit<Control<'a, OUT_N>>,
+    out_signal: Signal<(usize, [u8; OUT_N])>,
+}
+
+impl<'a, const IN_N: usize, const OUT_N: usize> State<'a, IN_N, OUT_N> {
+    pub fn new() -> Self {
+        State {
+            control: MaybeUninit::uninit(),
+            out_signal: Signal::new(),
+        }
+    }
+}
+
+pub struct HidClass<'d, D: Driver<'d>, const IN_N: usize, const OUT_N: usize> {
```

Comment:
The `IN_N` generic param is unused, it could be removed?

Reply:
It's also used to allocate a buffer on the stack for serializing input reports.

Reply:
oh, I was grepping for `IN_N` and didn't see it's called `N` in `ReportWriter`. Oops :rofl: